File: packages/server/src/routes/remote.ts
import type { FastifyInstance } from 'fastify';
import { getDb } from '../db.js';
import * as Q from '../queries.js';

export function registerRemoteRoutes(app: FastifyInstance): void {
  // ─── Node CRUD ──────────────────────────────────────

  app.get('/api/v1/remote/nodes', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const db = getDb();
    const nodes = Q.listRemoteNodes(db, userId);
    return { nodes };
  });

  app.post<{
    Body: { machine_name: string };
  }>('/api/v1/remote/nodes', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const { machine_name } = request.body ?? {};
    if (!machine_name || typeof machine_name !== 'string' || !machine_name.trim()) {
      return reply.status(400).send({ error: 'machine_name is required' });
    }

    const db = getDb();
    const node = Q.createRemoteNode(db, userId, machine_name.trim());
    return reply.status(201).send({ node });
  });

  app.delete<{
    Params: { id: string };
  }>('/api/v1/remote/nodes/:id', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const db = getDb();
    const node = Q.getRemoteNode(db, request.params.id);
    if (!node) return reply.status(404).send({ error: 'Node not found' });
    if (node.user_id !== userId) return reply.status(403).send({ error: 'Forbidden' });

    Q.deleteRemoteNode(db, node.id);
    return { ok: true };
  });

  // ─── Binding CRUD ───────────────────────────────────

  app.get<{
    Params: { nodeId: string };
  }>('/api/v1/remote/nodes/:nodeId/bindings', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const db = getDb();
    const node = Q.getRemoteNode(db, request.params.nodeId);
    if (!node) return reply.status(404).send({ error: 'Node not found' });
    if (node.user_id !== userId) return reply.status(403).send({ error: 'Forbidden' });

    const bindings = Q.listRemoteBindings(db, node.id);
    return { bindings };
  });

  app.post<{
    Params: { nodeId: string };
    Body: { channel_id: string; path: string; label?: string };
  }>('/api/v1/remote/nodes/:nodeId/bindings', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const db = getDb();
    const node = Q.getRemoteNode(db, request.params.nodeId);
    if (!node) return reply.status(404).send({ error: 'Node not found' });
    if (node.user_id !== userId) return reply.status(403).send({ error: 'Forbidden' });

    const { channel_id, path, label } = request.body ?? {};
    if (!channel_id || !path) {
      return reply.status(400).send({ error: 'channel_id and path are required' });
    }

    const channel = Q.getChannel(db, channel_id);
    if (!channel) return reply.status(404).send({ error: 'Channel not found' });

    const binding = Q.createRemoteBinding(db, node.id, channel_id, path, label ?? null);
    return reply.status(201).send({ binding });
  });

  app.delete<{
    Params: { nodeId: string; id: string };
  }>('/api/v1/remote/nodes/:nodeId/bindings/:id', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const db = getDb();
    const node = Q.getRemoteNode(db, request.params.nodeId);
    if (!node) return reply.status(404).send({ error: 'Node not found' });
    if (node.user_id !== userId) return reply.status(403).send({ error: 'Forbidden' });

    const binding = Q.getRemoteBinding(db, request.params.id);
    if (!binding || binding.node_id !== node.id) {
      return reply.status(404).send({ error: 'Binding not found' });
    }

    Q.deleteRemoteBinding(db, binding.id);
    return { ok: true };
  });

  // Channel-scoped binding list (owner only)
  app.get<{
    Params: { channelId: string };
  }>('/api/v1/channels/:channelId/remote-bindings', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const db = getDb();
    const bindings = Q.listChannelRemoteBindings(db, request.params.channelId, userId);
    return { bindings };
  });
}

File: packages/server-go/internal/api/remote.go
package api

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"

	"borgee-server/internal/auth"
	"borgee-server/internal/store"
)

type RemoteProxy interface {
	IsNodeOnline(nodeID string) bool
	ProxyRequest(nodeID string, action string, params map[string]string) (json.RawMessage, error)
}

type RemoteHandler struct {
	Store  *store.Store
	Logger *slog.Logger
	Hub    RemoteProxy
}

func (h *RemoteHandler) RegisterRoutes(mux *http.ServeMux, authMw func(http.Handler) http.Handler) {
	wrap := func(f http.HandlerFunc) http.Handler { return authMw(f) }

	mux.Handle("GET /api/v1/remote/nodes", wrap(h.handleListNodes))
	mux.Handle("POST /api/v1/remote/nodes", wrap(h.handleCreateNode))
	mux.Handle("DELETE /api/v1/remote/nodes/{id}", wrap(h.handleDeleteNode))
	mux.Handle("GET /api/v1/remote/nodes/{nodeId}/bindings", wrap(h.handleListBindings))
	mux.Handle("POST /api/v1/remote/nodes/{nodeId}/bindings", wrap(h.handleCreateBinding))
	mux.Handle("DELETE /api/v1/remote/nodes/{nodeId}/bindings/{id}", wrap(h.handleDeleteBinding))
	mux.Handle("GET /api/v1/channels/{channelId}/remote-bindings", wrap(h.handleListChannelBindings))
	mux.Handle("GET /api/v1/remote/nodes/{nodeId}/status", wrap(h.handleNodeStatus))
	mux.Handle("GET /api/v1/remote/nodes/{nodeId}/ls", wrap(h.handleNodeLs))
	mux.Handle("GET /api/v1/remote/nodes/{nodeId}/read", wrap(h.handleNodeRead))
}

func (h *RemoteHandler) handleListNodes(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodes, err := h.Store.ListRemoteNodes(user.ID)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to list nodes")
		return
	}
	if nodes == nil {
		nodes = []store.RemoteNode{}
	}
	writeJSONResponse(w, http.StatusOK, map[string]any{"nodes": nodes})
}

func (h *RemoteHandler) handleCreateNode(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	var body struct {
		MachineName string `json:"machine_name"`
	}
	if err := readJSON(r, &body); err != nil {
		writeJSONError(w, http.StatusBadRequest, err.Error())
		return
	}
	if body.MachineName == "" {
		writeJSONError(w, http.StatusBadRequest, "machine_name is required")
		return
	}

	node, err := h.Store.CreateRemoteNode(user.ID, body.MachineName)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to create node")
		return
	}

	writeJSONResponse(w, http.StatusCreated, map[string]any{"node": node})
}

func (h *RemoteHandler) handleDeleteNode(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	id := r.PathValue("id")
	node, err := h.Store.GetRemoteNode(id)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	if err := h.Store.DeleteRemoteNode(id); err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to delete node")
		return
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *RemoteHandler) handleListBindings(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodeID := r.PathValue("nodeId")
	node, err := h.Store.GetRemoteNode(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	bindings, err := h.Store.ListRemoteBindings(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to list bindings")
		return
	}
	if bindings == nil {
		bindings = []store.RemoteBinding{}
	}
	writeJSONResponse(w, http.StatusOK, map[string]any{"bindings": bindings})
}

func (h *RemoteHandler) handleCreateBinding(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodeID := r.PathValue("nodeId")
	node, err := h.Store.GetRemoteNode(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	var body struct {
		ChannelID string `json:"channel_id"`
		Path      string `json:"path"`
		Label     string `json:"label"`
	}
	if err := readJSON(r, &body); err != nil {
		writeJSONError(w, http.StatusBadRequest, err.Error())
		return
	}
	if body.ChannelID == "" || body.Path == "" {
		writeJSONError(w, http.StatusBadRequest, "channel_id and path are required")
		return
	}

	binding, err := h.Store.CreateRemoteBinding(nodeID, body.ChannelID, body.Path, body.Label)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to create binding")
		return
	}

	writeJSONResponse(w, http.StatusCreated, map[string]any{"binding": binding})
}

func (h *RemoteHandler) handleDeleteBinding(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodeID := r.PathValue("nodeId")
	node, err := h.Store.GetRemoteNode(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	bindingID := r.PathValue("id")
	if err := h.Store.DeleteRemoteBinding(bindingID); err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to delete binding")
		return
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *RemoteHandler) handleListChannelBindings(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	channelID := r.PathValue("channelId")
	bindings, err := h.Store.ListChannelRemoteBindings(channelID, user.ID)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to list bindings")
		return
	}
	if bindings == nil {
		bindings = []store.RemoteBinding{}
	}
	writeJSONResponse(w, http.StatusOK, map[string]any{"bindings": bindings})
}

func (h *RemoteHandler) handleNodeStatus(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodeID := r.PathValue("nodeId")
	node, err := h.Store.GetRemoteNode(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID && user.Role != "admin" {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	online := h.Hub != nil && h.Hub.IsNodeOnline(nodeID)
	writeJSONResponse(w, http.StatusOK, map[string]any{"online": online})
}

func (h *RemoteHandler) handleNodeLs(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodeID := r.PathValue("nodeId")
	node, err := h.Store.GetRemoteNode(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID && user.Role != "admin" {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	if h.Hub == nil || !h.Hub.IsNodeOnline(nodeID) {
		writeJSONError(w, http.StatusServiceUnavailable, "node_offline")
		return
	}

	path := r.URL.Query().Get("path")
	resp, err := h.Hub.ProxyRequest(nodeID, "ls", map[string]string{"path": path})
	if err != nil {
		if err == context.DeadlineExceeded {
			writeJSONError(w, http.StatusGatewayTimeout, "timeout")
			return
		}
		writeJSONError(w, http.StatusBadGateway, "Remote request failed")
		return
	}

	writeRemoteResponse(w, resp)
}

func (h *RemoteHandler) handleNodeRead(w http.ResponseWriter, r *http.Request) {
	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	nodeID := r.PathValue("nodeId")
	node, err := h.Store.GetRemoteNode(nodeID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Node not found")
		return
	}
	if node.UserID != user.ID && user.Role != "admin" {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	if h.Hub == nil || !h.Hub.IsNodeOnline(nodeID) {
		writeJSONError(w, http.StatusServiceUnavailable, "node_offline")
		return
	}

	path := r.URL.Query().Get("path")
	resp, err := h.Hub.ProxyRequest(nodeID, "read", map[string]string{"path": path})
	if err != nil {
		if err == context.DeadlineExceeded {
			writeJSONError(w, http.StatusGatewayTimeout, "timeout")
			return
		}
		writeJSONError(w, http.StatusBadGateway, "Remote request failed")
		return
	}

	writeRemoteResponse(w, resp)
}

func writeRemoteResponse(w http.ResponseWriter, resp json.RawMessage) {
	var parsed struct {
		Error string `json:"error"`
	}
	if err := json.Unmarshal(resp, &parsed); err == nil && parsed.Error != "" {
		switch parsed.Error {
		case "path_not_allowed":
			writeJSONError(w, http.StatusForbidden, parsed.Error)
		case "file_not_found":
			writeJSONError(w, http.StatusNotFound, parsed.Error)
		case "file_too_large":
			writeJSONError(w, http.StatusRequestEntityTooLarge, parsed.Error)
		case "timeout":
			writeJSONError(w, http.StatusGatewayTimeout, parsed.Error)
		default:
			writeJSONError(w, http.StatusBadGateway, parsed.Error)
		}
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	w.Write(resp)
}
